test/java/seedu/addressbook/data/AddressBookTest.java

```diff
@@ -0,0 +1,275 @@
+    @Test
+    public void syncTagsWithMasterList_emptyInitialTagList_addsAllIntoTagList() throws Exception {
+        UniqueTagList emptyTagList = new UniqueTagList();
+        UniquePersonList persons = new UniquePersonList(bobChaplin, charlieDouglas);
+
+        AddressBook addressBook = new AddressBook(persons, emptyTagList);
+
+        assertTrue(addressBook.containsTag(tagMathematician));
+        assertTrue(addressBook.containsTag(tagScientist));
+        assertFalse(addressBook.containsTag(tagEconomist));
+
+    }
+
+    @Test
+    public void syncTagsWithMasterList_someTagsNotInTagList_addsMissingTagsIntoTagList() throws Exception {
+        AddressBook addressBook = createDefaultAddressBook();
+
+        assertFalse(addressBook.containsTag(tagEconomist));
+        assertFalse(addressBook.containsTag(tagPrizeWinner));
+        addressBook.addPerson(davidElliot);
+        assertTrue(addressBook.containsTag(tagEconomist));
+        assertTrue(addressBook.containsTag(tagPrizeWinner));
+    }
+
+    @Test
+    public void syncTagsWithMasterList_defaultAddressBook_tagEqualityPropertyHolds() throws Exception {
+        AddressBook addressBook = createDefaultAddressBook();
+        UniquePersonList persons = addressBook.getAllPersons();
+        UniqueTagList tagsInPersons = getAllTags(persons);
+        UniqueTagList tags = addressBook.getAllTags();
+        
+        // all person tags are equal to some tag in the master list
+        boolean allPersonTagsValid = true;
+        for (Tag personTag : tagsInPersons) {
+            // particular person tag is equal to some tag in the master list
+            boolean personTagValid = false;
+            for (Tag tag : tags) {
+                personTagValid = personTagValid || (personTag == tag);
+            }
+            allPersonTagsValid = allPersonTagsValid && personTagValid;
+        }
+        
+        assertTrue(allPersonTagsValid);
+    }
+
+    @Test
+    public void addPerson_personNotInList_addsNormally() throws Exception {
+        AddressBook addressBook = createDefaultAddressBook();
+        
+        // add person whose tag is in tag list
+        assertFalse(addressBook.containsPerson(charlieDouglas));
+        addressBook.addPerson(charlieDouglas);
+        assertTrue(addressBook.containsPerson(charlieDouglas));
+        
+        // add person whose tag is not in tag list
+        assertFalse(addressBook.containsPerson(davidElliot));
+        addressBook.addPerson(davidElliot);
+        assertTrue(addressBook.containsPerson(davidElliot));
+    }
+
+    @Test
+    public void addPerson_personAlreadyInList_throwsDuplicatePersonException() throws Exception {
+        AddressBook addressBook = createDefaultAddressBook();
+        
+        thrown.expect(DuplicatePersonException.class);
+        addressBook.addPerson(aliceBetsy);
+    }
+
+    @Test
+    public void addTag_tagNotInList_addsNormally() throws Exception {
+        AddressBook addressBook = createDefaultAddressBook();
```

Comment:
Looks like this line appears in every `@Test`. May be put that in the `@Before` method?